### ISSUE TITLE
Optional path params, or workarounds for them.

### ISSUE THREAD

**Issue by threehams** [typehero/type-route#24]
Optional path params aren't ideal, but sometimes they happen. At my company (and others) any SEO-critical route will be controlled by the SEO team, not by development. So, that leads to things like this real-world example: [this nightmare of a URL](https://www.truecar.com/new-cars-for-sale/listings/acura/ilx/year-2017-2020/body-sedan/price-12000-88000/?driveTrain[]=FWD&interiorColorGeneric[]=Black&interiorColorGeneric[]=Brown), where everything past `/listings/` is optional.

I'm not asking for type-route to support this URL exactly (nothing besides custom code can), but I wonder if there's a way to support some kind of partial or untyped escape-hatch solution.

## Optional path params (with... complications)
The above URL could possibly be represented by this string, but only if evaluated either right to left, or by
`/<<<newness>>>-cars-for-sale/listings/<<<make?>>>/<<<model?>>>/year-<<<years?>>>/body-<<<body?>>>/price-<<<price?>>>/`

This gets complicated. The optional part is the entire fragment with `//` boundaries, which no router I know of supports, since they're using `path-to-regexp`. `type-route` doesn't currently support mixing params with literals (maybe another issue).

## ** escape hatch

These URLs don't really happen much, so just being able to wildcard the whole route would work fine.
`/<<<newness>>>/listings/**`

Thoughts on these?

**Comment by bradenhs** [comment]
Thanks for bringing this use case up! Though obviously not ideal you could still technically support this scenario by using the underlying history object Type Route exposes (https://www.type-route.org/docs/api-reference/router/history). For matching you'd just check the current url doesn't match any Type Route routes and then use custom code to determine if a page with a url not expressible using Type Route should be displayed before falling back to a not found page. I would like to support this in a first class way if possible but need to think on exactly how it would work a little bit more.

**Comment by bradenhs** [comment]
This PR addresses this use case: https://github.com/type-route/type-route/pull/35. Any feedback on the updated api around trailing and optional path parameters would be appreciated.

**Comment by threehams** [comment]
Hi! Sorry for not responding yet - been really busy with non-routing-related priorities at work, and I've started and continued a game jam project in free time (which is a single route, otherwise I'd be using `@next` already!).

This is a lot to take in. I'm really glad you managed to find a way to deal with the `{ parse, stringify }` by convention, given the limitations of dependent generics in TS. I tested out some of the crazier routes in [this fork of the playground](https://codesandbox.io/s/splat-routes-vrxoh?file=/src/playground.tsx) and kept failing to find cracks in it. It's almost magic the way everything's coming together, especially with the addition of `default()`.

What's your hopeful timeline for release? I'd like to schedule in an attempt to integrate this into our code in the future, and present it to the team, but have to work around priorities from COVID-19.

**Comment by bradenhs** [comment]
A couple big things need to be done before release (1) fill out the unit/integration test suite - this update include major code changes under the hood and I want to be more confident in those changes before releasing it (2) update the documentation. Its hard to put a timeline on when those 2 things will be done. I'm definitely going to prioritize polish over timely delivery of this. That being said, if things go smoothly, maybe a month? That's a wild guess though, could be sooner or later than that. In the meantime any feedback on the api or any bug reports would be a huge help!

**Comment by bradenhs** [comment]
Released support for this in `0.4.0`! Feel free to reopen if you feel that release doesn't address this fully.
